californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
I may not have fully understood the overall idea here but here's how I think it *should* work.
1. client code creates an instance of `CoapClient` and starts observing a resource using e.g. `CoapClient.observeAndWait(CoapHandler, int)`
1. `CoapClient` creates an `ObserveMessageObserverImpl` as already done in the original code and registers it with the original `Request` object.
1. the original request is put to ÒbservationStore`
1. when a notification comes in the original request is retrieved from the `ObservationStore` and the `MessageObserver.onResponse()` method is invoked (as part of `Request.setResponse()`)
1. the *global* `NotificationListener` registered on the endpoint is also notified about the response resceived in reply to the observe request.

I do not see any need for registering the additional observer specific `NotificationListener`adapter. In case the node on which the original observe request (including the handler object) has been issued crashes, there is no longer a handler instance to be notified anyways, or is there?
I think client code that is supposed to also work in case of a fail-over, must not use the *individual message observer* approach for observing resources but must instead register a notification listener at the endpoint level from the beginning.
In that regard I do not think that we are changing the behavior of `CoapClient` but instead only add support for fail-over scenarios by means of new (alternative) API.